src/contract/assembled_transaction.ts

```diff
@@ -411,21 +429,70 @@ export class AssembledTransaction<T> {
     return tx;
   }
 
-  simulate = async (): Promise<this> => {
+  private static async buildFootprintRestoreTransaction<T>(
```

Comment:
Should this be part of the exposed API of `AssembledTransaction`? Seems like it might be too low-level to me. What with it taking `sorobanData` as an argument (the test makes me think that constructing this `sorobanData` is [not straightforward](https://github.com/stellar/js-stellar-sdk/pull/974/files#diff-5835ad6a083d1becead94519935fb7ac135df05ce4df4ac04584f0fc664f5614R13)).

If it _should_ be part of the public API, then it needs good documentation. And maybe that `sorobanData` argument can get rid of the `this | that` intersection type, and just take the easier-to-construct variant for now, to start with as limited a public API as possible.

If you think maybe it should not be, where else do you think it could live? 

- A non-exposed utility function, that just lives inside this file? 
- An exposed utility function, exposed in the `contract.*` import? (Like `import { buildFootprintRestoreTransaction } from '@stellar/stellar-base/contract'`)
- An exposed utility function, exposed somewhere else?

Reply:
Oh, I see. Actually hadn't noticed that `buildFootprintRestoreTransaction` is marked private. 🤦🏻

Still not sure why it's part of AssembledTransaction, though! Seems like it could easily just be a non-exported function in this same file.

As for `updateTimeout: false` and `force: false`—I wonder if it would be better to make the spots that caused problems deal more gracefully with restore transactions. Like:

```ts
if (!force && !this.isRestore && this.isReadCall) {
  throw new AssembledTransaction.Errors.NoSignatureNeeded(
    "This is a read call. It requires no signature or sending. " +
    "Use `force: true` to sign and send anyway."
  );
}
```

And as for the error you were getting when updating the timeout, it seems like maybe we should move the `if (Api.isSimulationRestore(simulation)) { throw new AssembledTransaction.Errors.ExpiredState(...) }` to `get result`, instead of `get simulationData`? It seems like maybe we actually can construct both pieces of `simulationData` correctly, when given a restore transaction. Looking at the definition of `isSimulationRestore` leads to:

```ts
  export interface SimulateTransactionRestoreResponse
    extends SimulateTransactionSuccessResponse {
    result: SimulateHostFunctionResult; // not optional now
```

Which makes it seem like the `get simulationData` should work just fine for restore transactions. We only need the `result` and `transactionData` keys:

```ts
this.simulationResult = simulation.result;
this.simulationTransactionData = simulation.transactionData.build();
```

So yeah, I think we can move that `if (Api.isSimulationRestore(simulation))` check to `get result`, because _that's_ the one that will cause problems (presumably) if you didn't restore it yet. And then you can get rid of the `updateTimeout` logic.

Reply:
The problem with `get simulationData` on the `restoreFootprint` transaction is not the `API.isSimulationRestore` check, its the last check: `if(!simulation.result)` - there is no result for such a transaction apparently (I'm not sure why)

Reply:
Very strange, since it's typed specifically in a way to say that `result` will _always_ be there, for a `SimulateTransactionRestoreResponse`, as I showed above. 

Here's the logic in question:

```ts
    if (!simulation.result) {
      throw new Error(
        `Expected an invocation simulation, but got no 'result' field. Simulation: ${JSON.stringify(
          simulation,
          null,
          2,
        )}`,
      );
    }

    // add to object for serialization & deserialization
    this.simulationResult = simulation.result;
    this.simulationTransactionData = simulation.transactionData.build();

    return {
      result: this.simulationResult,
      transactionData: this.simulationTransactionData!,
    };
```

I wonder if we should just stop caring. Remove the `if (!simulation.result)` check. It'll set `simulationData.result` to whatever empty value. That's fine?

Then we'll need to update `get result` a little. Probably still don't need the big `if (!result)` error logic. This might be enough?

```diff
 get result(): T {
   try {
-    return this.options.parseResultXdr(this.simulationData.result.retval);
+    return this.options.parseResultXdr(this.simulationData.result?.retval);
   } catch (e) {
     if (!implementsToString(e)) throw e;
     const err = this.parseError(e.toString());
     if (err) return err as T;
     throw e; // eslint-disable-line
   }
 }
```

Then if `result` was empty, well... `result` will be empty. Not sure we need to shout about that!

Reply:
@chadoh I implemented your suggestions - removed the `if (!simulation.result)` check, updated the types to have ` | undefined`, added the `?` needed - all tests pass except this one: 
```
  custom-types › u32_fail_on_even

  test/e2e/src/test-custom-types.js:34

   33:   t.deepEqual(                                                    
   34:     (await t.context.client.u32_fail_on_even({ u32_: 2 })).result,
   35:     new contract.Err({ message: "Please provide an odd number" }),

  Rejected promise returned by test. Reason:

  Error {
    message: `Transaction simulation failed: "HostError: Error(Contract, #1)␊
    ␊
  ```

Might just need to update the error message, but I'm not sure what this test is supposed to be checking. Any thoughts?

Reply:
Weird! That doesn't seem like an error caused by the removal of `if (!simulation.result)`. This test is checking that contract functions that return a Rust `Result` type get encoded as a similar `Result` type in the JS, too. So here we're _expecting_ it to throw an error, but we're expecting it to get caught by that `catch` block, and then the `parseError` function should see that `HostError` and turn it into the `contract.Err` with the message provided by the contract author ("Please provide an odd number").

Reply:
I also don't see the changes you described. The `if (!simulation.result)` is still there, in the `get simulationData` function. Did you forget to push?

Reply:
I didn't push it yet with the failing test, was planning to figure out what was up with that first. I'll dig in to see why its not getting caught by the catch / parseError function.

Reply:
Got it working with the latest commit. I had to add a check inside the try catch block of `result`. The check needs to be inside so that when it tries to access `simulationData` it will fail with the `SimulationError` which is then parsed with `parseError` (which then returns the correct error for the test).